File: src/leetcode/java/ConvertBSTToGreaterTree_538.java
package leetcode.java;

public class ConvertBSTToGreaterTree_538 {
  public TreeNode convertBST(TreeNode root) {
    if (root == null) return null;
    int[] res = new int[1];
    bstTrav(root, res);
    return root;
  }

  private void bstTrav(TreeNode node, int[] res) {
    if (node == null) return;
    bstTrav(node.right, res);
    int tmp = node.val;
    node.val += res[0];
    res[0] += tmp;
    bstTrav(node.left, res);
  }

  class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode(int x) {
      val = x;
    }
  }
}

File: src/leetcode/python/ConvertBSTToGreaterTree_538.py
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class ConvertBSTToGreaterTree_538(object):
    def convertBST(self, root):
        """
        :type root: TreeNode
        :rtype: TreeNode
        """
        if not root:
            return None
        res = []

        def bstTrav(node):
            if not node:
                return
            bstTrav(node.right)
            tmp = node.val
            node.val += res[0]
            res[0] += tmp
            bstTrav(node.left)
            return

        bstTrav(root)
        return root
